### DISCUSSION TITLE
AI perks reduction

### DISCUSSION THREAD

**Discussion by ihhub** [ihhub/fheroes2#5797] [General]
Since our AI has become more superior over the original AI in many areas many players express concern about the perks AI has. These privileges were given to AI players in order to make it more competitive but in fheroes2 it makes it extremely difficult to beat under certain conditions.
My suggestion to reduce the amount of perks AI has. This is by far the full list of perks AI has:
1. Easy difficulty: gold income is reduced by 25%.
2. Normal difficulty: scouting area +1.
3. Hard difficulty: scouting area +2, gold income is increased by 29%, unit growth +20%.
4. Expert difficulty: scouting area +3, gold income is increased by 45%, unit growth +32%, heroes have extra 75 points for movement.
5. Impossible difficulty: scouting area +4, gold income is increased by 60%, unit growth +44%, heroes have extra 75 points for movement.

My biggest concern is unit growth which can be abused by human players as well. We do not need to remove all perks but to reduce them.

**Comment by Alucard648** [comment]
I think, capturing enemy castle on Hard or higher should kill all creatures in castle dwellings, leaving player with no one to recruit from.

**Reply by Branikolog** [reply]
Even in a case AI accidentally captured your castle full of Black Dragons, and you captured it back on the next day?)

**Reply by Alucard648** [reply]
Yes. And on Expert/Impossible those dragons will automatically join enemy army for 50%discount/free.

**Reply by Branikolog** [reply]
In my mind the game should be as much possible fair, as it could, considering all the weaknesses of AI. I don't think it's a good idea providing AI some ultimate perks, which couldn't be achieved in any circumstances by human player. 
I believe, that in ideal game AI should play with 100% the same conditions as human player has. ) Or we should at least strive for it. ;)

**Comment by LeHerosInconnu** [comment]
Hello @ihhub and @Alucard648,

[Here](https://www.celestialheavens.com/strategy-guide) is an article with information about the difficulty levels of the original game.

Easy:
- "AI starts with 7,500 gold, 20 wood and ore, and 5 of each rare resource."
- "AI will occasionally do some really stupid things, such as attacking a hero or castle with no real chance of winning."
- "I think the AI is prevented from purchasing level 6 creatures if it built the dwelling itself."

Normal:
- "AI starts with 10,000 gold, 30 wood and ore, and 10 of each rare resource."
- "As far as I know, the level 6 creature handicap is still in effect."

**Edit:**
At Normal difficulty AI has the same bonus as for Easy difficulty concerning gold and resources:
- "AI starts with 7,500 gold, 20 wood and ore, and 5 of each rare resource."

Hard:
- "AI starts with 10,000 gold, 30 wood and ore, and 10 of each rare resource."
- "AI players receive a 10% increase in gold income (i.e., their castles produce 1,100 gold per day) and get one free wood and ore each turn."
- "I know for sure that the level 6 creature handicap is gone."
- "Computer heroes will only attack if they have at least a 50% advantage in total hit points."

Expert:
- "AI starts with 10,000 gold, 30 wood and ore, and 10 of each rare resource."
- "AI gains 50% more gold and gets one of every resource free each turn, even if he doesn't own any mines."
- "Computer heroes like to have an army three times larger than yours before considering an attack."

Impossible:
- "AI starts with 10,000 gold, 30 wood and ore, and 10 of each rare resource."
- "Furthermore, they gain two of each resource each turn, free of charge, and they make twice as much gold as you would with the same amount of towns, castles, and mines."
- "AI will wait for a 900% hit point advantage before attacking."

More Computer Advantages:
- "AI also does not seem to be limited by black, unexplored area."
- "AI also appears to have perfect information on every hero and every creature stack."

We don't have to follow the original game to the letter as far as AI bonuses are concerned.

For me, the scouting and movement bonuses are fine (in fheroes2 the AI doesn't ignore the fog of war, if that is correct).
Gold and resources bonuses should maybe be increased for the AI, at least with the same values as the original game.
The creature growth bonus should simply be removed, as the human player can benefit from it indirectly.

I think the AI should have enough gold and resources to recruit most of the creatures available in its cities.

Some of the "official" scenarios and others proposed by scenario creators were created with the Heroes 2 AI in mind (even more so the scenarios created for the "The Price of Loyalty" expansion).
I don't think that there is a universal solution to solve the problems related to the differences between the AI of the original game and the AI of fheroes2 for all the scenarios already created, unless you specifically modify the behavior of the AI of fheroes2 for these scenarios or make corrections to these scenarios in the scenario editor to get a corrected version adapted for fheroes2.

> I think, capturing enemy castle on Hard or higher should kill all creatures in castle dwellings, leaving player with no one to recruit from.

I don't think this is a good idea.
In my opinion, as I wrote above, it would be better if the AI always had enough gold and resources to recruit the majority of the creatures available in its cities (up to a certain point of course, so that the human player in a position of dominance over the AI could still have fun recruiting creatures from captured opponent cities). :)

**Comment by ihhub** [comment]
This is interesting. AI in fheroes2 has only 10000 gold, 30 wood and ore and, 10 of sulfur, gems and crystals no matter what difficulty we choose. This is definitely a bug.

**Comment by Branikolog** [comment]
The less extra creatures AI gets - the fairer gameplay would be. :)

**Comment by zenseii** [comment]
Hi, guys.

There are many interesting points here and I have some thoughts as well after having looked at the PR #5900 .

The way I see it, the bonuses the AI receives are there to compensate for *specific* shortcomings it has.


1. Gold bonus.
- The AI wastes lots of gold by sending its heroes to die, particularly by ignoring enemy hero threat area, and this leads to them often having to recruit heroes over again, so this bonus is fair as long as the AI wastes a lot of gold on this (Yes, many human players don't play much better.)

2. Resources.
- The AI doesn't know how to use the marketplace, so the human player has an advantage that needs to be compensated for here. EDIT: The AI has learnt to use the Marketplace for town construction.

3. Creature growth.
- I'm not too sure about this one and why it is needed for the AI, as it seems to be quite good at not wasting creatures in battle. Edit: I've found that autoresolve often loses some creatures when you have rangers and fast walkers like boars, against slow walkers. However, the losses aren't so big that they should get a bonus imo.
- It does tend to attack every neutral army it finds, whether it guards something important or not. This, coupled with the above point, means that they lose more than they should.

4. Scouting
- After using the AI control feature, it has become obvious to me that the AI will scour every small corner of the map systematically to uncover as much of it as possible. However, less or more scouting doesn't really affect this, since they would still be searching out every corner of the map.
- More scouting range does help the AI locate things that will pull it out of the "scouting mode" faster though.

5. Movement points
- Heroes 2 has logistics and planning of your movement as a key part of the gameplay, like walking over adventure map objects to cut a route short. The AI is better than the average player at this. However, the main reason for their movement point boost is that the AI will carry slow troops like peasants and dwarves around, which will only slow them down, and by a lot. Novice players make the same mistake, but more experienced players don't.
- Extra movement points is a decent compensation for this shortcoming until it gets sorted out.

**Comment by ihhub** [comment]
**Update**:
We decreased monster growth bonus for AI:
- Easy - 1.0 (no bonus)
- Normal - 1.0 (no bonus)
- Hard - 1.15 (was 1.20)
- Expert - 1.26 (was 1.32)
- Impossible - 1.37 (was 1.44)

**Comment by zenseii** [comment]
Here's a relevant previous discussion about the AI perks in both the original game and fheroes2: https://github.com/ihhub/fheroes2/issues/657
Note that it is quite old and some opinions about the goals of the project might have changed.

It is interesting to see that the article you linked @LeHerosInconnu has wrong info about the original AI bonuses with regards to the gold boost for impossible difficulty.
